File: cpp/1106_解析布尔表达式.cpp
/*
 * @lc app=leetcode.cn id=1106 lang=cpp
 *
 * [1106] 解析布尔表达式
 */

#include "AllInclude.h"

// @lc code=start
/**
 * Accepted 2022-11-05
 * 75/75 (8 ms)
 * Your runtime beats 58.10 % of cpp submissions
 * Your memory usage beats 98.00 % of cpp submissions (6.7 MB)
 */

class Solution {
public:
    bool parseBoolExpr(string expression) { return help(expression, 0, expression.length()); }

private:
    bool help(string& str, size_t begin, size_t end) {
        size_t i = begin + 2;
        size_t j = i + 1;
        uint16_t stack = 0;
        switch (str[begin]) {
            case 'f':
                return false;
                break;

            case 't':
                return true;
                break;

            case '!':
                return !help(str, i, end - 1);
                break;

            case '&':
                while (j < end) {
                    if (str[j] == '(') {
                        ++stack;
                    }

                    if (str[j] == ')') {
                        --stack;
                    }

                    if (str[j] == ',' && stack == 0) {
                        // 遇到一个表达式
                        // 短路运算
                        if (!help(str, i, j)) {
                            return false;
                        }
                        i = j + 1;
                    }
                    ++j;
                }
                return help(str, i, j - 1);
                break;

            default:  // |
                while (j < end) {
                    if (str[j] == '(') {
                        ++stack;
                    }

                    if (str[j] == ')') {
                        --stack;
                    }

                    if (str[j] == ',' && stack == 0) {
                        // 遇到一个表达式
                        // cout << "i: " << i << " j: " << j << "\n";
                        // 短路运算
                        if (help(str, i, j)) {
                            return true;
                        }
                        i = j + 1;
                    }
                    ++j;
                }
                // cout << "i: " << i << " j: " << j << "\n";
                return help(str, i, j - 1);
                break;
        }
    }
};
// @lc code=end
/*
""!(f)"\n
"|(f,t)"\n
"&(t,f)"\n
"|(&(t,f,t),!(t))"\n
"|(&(|(f)),|(f,f,f,t),&(t,f,t),!(t))"\n
"!(&(&(!(&(f)),&(t),|(f,f,t)),&(t),&(t,t,f)))""
*/

File: leetcode_practice/src/1106_解析布尔表达式.rs
/*
 * @lc app=leetcode.cn id=1106 lang=rust
 *
 * [1106] 解析布尔表达式
 */

// @lc code=start
/*
    Accepted 2022-11-05
    75/75 (0 ms)
    Your runtime beats 100 % of rust submissions
    Your memory usage beats 50 % of rust submissions (2 MB)
*/
impl Solution {
    pub fn parse_bool_expr(expression: String) -> bool {
        Solution::help(&expression[..])
    }

    fn help(expression: &str) -> bool {
        let mut char_iter = expression.chars();

        if let Some(c) = char_iter.nth(0) {
            let mut i = 2;
            let mut j = i + 1;
            let mut stack = 0;
            match c {
                'f' => false,

                't' => true,

                '!' => !Solution::help(&expression[i..(expression.len() - 1)]),

                '&' => {
                    char_iter.nth(1);
                    while j < expression.len() {
                        if let Some(c_j) = char_iter.nth(0) {
                            if c_j == '(' {
                                stack += 1;
                            } else if c_j == ')' {
                                stack -= 1;
                            } else if c_j == ',' && stack == 0 {
                                if !Solution::help(&expression[i..j]) {
                                    return false;
                                }

                                i = j + 1;
                            }
                        }
                        j += 1;
                    }
                    Solution::help(&expression[i..(j - 1)])
                }

                _ => {
                    char_iter.nth(1);
                    while j < expression.len() {
                        if let Some(c_j) = char_iter.nth(0) {
                            if c_j == '(' {
                                stack += 1;
                            } else if c_j == ')' {
                                stack -= 1;
                            } else if c_j == ',' && stack == 0 {
                                if Solution::help(&expression[i..j]) {
                                    return true;
                                }

                                i = j + 1;
                            }
                        }
                        j += 1;
                    }
                    Solution::help(&expression[i..(j - 1)])
                }
            }
        } else {
            false
        }
    }
}
// @lc code=end

